### PR TITLE
feat: create direct controller for DataflowFlexTemplateJob

### PR DESCRIPTION
**Pull request by justinsb**

### CODE REVIEW

File: pkg/controller/direct/dataflow/refs.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+	if subnetworkRef, ok := v.(*refs.ComputeSubnetworkRef); ok {
+		resolved, err := refs.ResolveComputeSubnetwork(r.ctx, r.kube, r.src, subnetworkRef)
+		if err != nil {
+			return err
+		}
+		*subnetworkRef = *resolved
+	}
+
+	if subnetworkRefs, ok := v.([]refs.ComputeSubnetworkRef); ok {
```
Comment (yuwenma) [inline pkg/controller/direct/dataflow/refs.go:73]: Cool!  didn't know we can do this type assertion in go!

**Review by yuwenma — COMMENTED** [review]
/lgtm
/approve

/hold one minor about the errors. Feel free to unhold.

File: pkg/controller/direct/dataflow/refs.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		for i := range subnetworkRefs {
+			subnetworkRef := &subnetworkRefs[i]
+			resolved, err := refs.ResolveComputeSubnetwork(r.ctx, r.kube, r.src, subnetworkRef)
+			if err != nil {
+				return err
+			}
```
[the remark this answers is not in the document]
Comment (justinsb) [reply pkg/controller/direct/dataflow/refs.go:79]: Honestly I don't know.  Resolving each one could be a k8s GET, which could add up, but OTOH I wouldn't expect too many errors here.

I do want us to move this to a shared ref-resolver (I feel like there's a PR somewhere, but I don't know where...).  Maybe we tackle this at that time?

[the remark this answers is not in the document]
Comment (yuwenma) [reply pkg/controller/direct/dataflow/refs.go:79]: 👌

**Pull request merged**